### PR TITLE
Update hdf5 and fms versions for spack-stack/1.4.0

### LINKED ISSUE #524: mapl 2.35.2 build errors on Narwhal
**Describe the bug**
mapl@2.22.0 builds fine on Narwhal (a Cray) with Intel, but 2.35.2 chokes with:
```
     784     Running Python2 import test on ShaveMantissa_
     785     cd /p/app/projects/NEPTUNE/spack-stack/spack-stack-1.3.0/cache/build_stage/heinzell/spack-stage-mapl-2.35.2-24y4r3s7yvhietdiewh2jv2o7c6lvnut/spack-build-24y4r3s/MAPL_cfio_r4 && /p/app/projects/NEPTUNE/spack-stack/spac
             k-stack-1.3.0/envs/unified-env-intel-2021.4.0/install/intel/2021.4.0/cmake-3.23.1-tfol62l/bin/cmake -E env LD_LIBRARY_PATH=/p/app/projects/NEPTUNE/spack-stack/spack-stack-1.3.0/cache/build_stage/heinzell/spack-stage-m
             apl-2.35.2-24y4r3s7yvhietdiewh2jv2o7c6lvnut/spack-build-24y4r3s/lib:/opt/cray/pe/libsci/22.08.1.1/INTEL/19.0/x86_64/lib:/opt/cray/pe/gcc/10.3.0/snos/lib:/opt/cray/pe/gcc/10.3.0/snos/lib64:/opt/intel/oneapi_2021.4.0.34
             22/mkl/2021.4.0/lib/intel64:/opt/intel/oneapi_2021.4.0.3422/compiler/2021.4.0/linux/lib:/opt/intel/oneapi_2021.4.0.3422/compiler/2021.4.0/linux/lib/x64:/opt/intel/oneapi_2021.4.0.3422/compiler/2021.4.0/linux/lib/emu:/
             opt/intel/oneapi_2021.4.0.3422/compiler/2021.4.0/linux/lib/oclfpga/host/linux64/lib:/opt/intel/oneapi_2021.4.0.3422/compiler/2021.4.0/linux/lib/oclfpga/linux64/lib:/opt/intel/oneapi_2021.4.0.3422/compiler/2021.4.0/lin
             ux/compiler/lib/intel64_lin:/opt/cray/pe/papi/6.0.0.16/lib64:/opt/cray/libfabric/1.11.0.4.125/lib64 Python2_EXECUTABLE-NOTFOUND -c import\ ShaveMantissa_
     786     No such file or directory
  >> 787     make[2]: *** [MAPL_cfio_r4/CMakeFiles/ShaveMantissa_.dir/build.make:84: ShaveMantissa_] Error 1
     788     make[2]: Leaving directory '/p/app/projects/NEPTUNE/spack-stack/spack-stack-1.3.0/cache/build_stage/heinzell/spack-stage-mapl-2.35.2-24y4r3s7yvhietdiewh2jv2o7c6lvnut/spack-build-24y4r3s'
  >> 789     make[1]: *** [CMakeFiles/Makefile2:652: MAPL_cfio_r4/CMakeFiles/ShaveMantissa_.dir/all] Error 2
     790     make[1]: *** Waiting for unfinished jobs....
```
I don't know why it thinks that it needs to have Python 2. The spack `package.py` doesn't list Python as a dependency. And if it wants Python, then it needs to asl for Python 3 (and not 2) in the spack package recipe.

**To Reproduce**
Try to build mapl@2.35.2 on Narwhal with the existing Intel (or GNU) site config.

**Expected behavior**
It builds!

**System:**
Narwhal, Intel 2021.4.0

**Additional context**
n/a

### LINKED ISSUE #553: mapl@2.35.2 build errors on macOS after esmf@8.4.1 build errors are fixed in 8.4.2b02
**Describe the bug**
```
  >> 1769    Undefined symbols for architecture x86_64:
     1770      "__ZNKSt13runtime_error4whatEv", referenced from:
     1771          __ZNK8nlohmann16json_abi_v3_11_26detail9exception4whatEv in libesmf.a(ESMC_InfoCDef.o)
     1772          __ZNK8nlohmann16json_abi_v3_11_26detail9exception4whatEv in libesmf.a(ESMC_InfoCDefGeneric.o)
     1773          __ZNK8nlohmann16json_abi_v3_11_26detail9exception4whatEv in libesmf.a(ESMCI_Array.o)
     1774          __ZNK8nlohmann16json_abi_v3_11_26detail9exception4whatEv in libesmf.a(ESMCI_Base.o)
     1775          __ZNK8nlohmann16json_abi_v3_11_26detail9exception4whatEv in libesmf.a(ESMCI_Field_F.o)

     ...

     3079          _c_esmc_arrayread_ in libesmf.a(ESMCI_Array_F.o)
     3080          _c_esmc_arraysmmstoreind4_ in libesmf.a(ESMCI_Array_F.o)
     3081          _c_esmc_arraysmmstoreind8_ in libesmf.a(ESMCI_Array_F.o)
     3082          _c_esmc_arraysmmstorenf_ in libesmf.a(ESMCI_Array_F.o)
     3083          ...
     3084    ld: symbol(s) not found for architecture x86_64
  >> 3085    collect2: error: ld returned 1 exit status
  >> 3086    make[2]: *** [bin/cub2latlon.x] Error 1
  >> 3087    make[1]: *** [base/CMakeFiles/cub2latlon.x.dir/all] Error 2
     3088    make[1]: *** Waiting for unfinished jobs....
     3089    /usr/local/Cellar/cmake/3.23.1/bin/cmake -E cmake_copy_f90_mod include/MAPL.generic/mapl_statespecification.mod generic/CMakeFiles/MAPL.generic.dir/mapl_statespecification.mod.stamp GNU
     3090    /usr/local/Cellar/cmake/3.23.1/bin/cmake -E touch generic/CMakeFiles/MAPL.generic.dir/StateSpecification.F90.o.provides.build
     3091    [ 91%] Building Fortran object generic/CMakeFiles/MAPL.generic.dir/ComponentSpecification.F90.o
```
The missing symbols error in esmf 8.4.1, fixed in 8.4.2b02, seems to be handed down to mapl.

**To Reproduce**
Build mapl@2.35.2 with esmf@8.4.2b02

**Expected behavior**
It builds

**System:**
macOS with clang + gfortran

**Additional context**
n/a

### LINKED ISSUE #586: Update hdf5 & fms for spack-stack/1.4.0
**Is your feature request related to a problem? Please describe.**
The UFS-WM will soon upgrade to `hdf5/1.14.0` and  `fms/2023.01` (see [UFS-WM #1745](https://github.com/ufs-community/ufs-weather-model/pull/1745)). `spack-stack/1.3.0` used `hdf5/1.14.0`, but the release of `spack-stack/1.3.1` reverted to `hdf5/1.12.2` due to a Skylab issue (see #546). 

Just a note: also to-be-updated in the aforementioned UFS-WM PR are `netcdf` (4.74 --> 4.9.2), `esmf` (8.3.0b09 --> 8.4.2), and `mapl` (2.22.0 --> 2.35.2). We already include `netcdf-c/4.9.2` by default, while `esmf/8.4.2` and `mapl/2.35.2` are appended to the unified-env installations.

**Describe the solution you'd like**
`hdf5/1.14.1` is now available, so we should switch to `hdf5/1.14.1` and `fms/2023.01` for the `spack-stack/1.4.0` release.

### PR DESCRIPTION
**Pull request by ulmononian**
## Description

Updates `hdf5`, `fms`, `esmf`, and `mapl` versions (to 1.14.1, 2023.01, 8.4.2, and 2.35.2, respectively) in anticipation of forthcoming ufs-wm changes (see #586, [UFS-WM #1745](https://github.com/ufs-community/ufs-weather-model/pull/1745)). These versions should be used in the `spack-stack/1.4.0` release.

## Definition of Done
`hdf5`, `fms`, `esmf`, and `mapl` versions are updated in configs/common/packages.yaml.

### Issue(s) addressed
Fixes #586 
Fixes https://github.com/JCSDA/spack-stack/issues/553
Fixes https://github.com/JCSDA/spack-stack/issues/524

## Dependencies
[spack #268](https://github.com/JCSDA/spack/pull/268)

## Impact
None known

### CODE REVIEW

**Comment by climbfuji** [comment]
We'll need a PR for spack that adds those versions, in case they don't exist yet. Also, please bump `py-h5py` back to 3.7.0 (3.6.0 is not compatible with hdf5 1.14.x).

**Comment by climbfuji** [comment]
I am also in favor of updating to esmf@8.4.2 and mapl@2.35.2, this will solve a whole lot of problems on Narwhal, Gaea, macOS, ...

**Comment by ulmononian** [comment]
> We'll need a PR for spack that adds those versions, in case they don't exist yet. Also, please bump `py-h5py` back to 3.7.0 (3.6.0 is not compatible with hdf5 1.14.x).

done (https://github.com/JCSDA/spack/pull/268) & done.

**Comment by ulmononian** [comment]
> I am also in favor of updating to esmf@8.4.2 and mapl@2.35.2, this will solve a whole lot of problems on Narwhal, Gaea, macOS, ...

i updated the default esmf/mapl versions in common/packages.yaml.

File: configs/common/packages.yaml
Comment (AlexanderRichert-NOAA) [inline configs/common/packages.yaml]: These variants need to get updated. `snapshot` should be `none`, and I believe we want to move to `+external-parallelio` (with no `esmf_pio` setting).

Comment (AlexanderRichert-NOAA) [reply configs/common/packages.yaml]: Also I'm trying to remember the story behind `~shared`. That was for UFS WM, right? Might be good to double check whether we really need it.

Comment (ulmononian) [reply configs/common/packages.yaml]: @AlexanderRichert-NOAA thanks for taking a look. i'll change `snapshot` to `none`. can we get confirmation from @climbfuji regarding removing the `esmf_pio` setting and adding `+external-parallelio`?

i believe `~shared` was included due to those ufs-wm issues we were having a couple months ago. right now, it indeed looks like all the esmf installations from `spack-stack/1.3.0` onward were done with only static libraries. building again w/ shared might bring back those issues and i'd prefer to avoid that as we are close to getting the ufs-wm modulefile pr merged. thoughts?

Comment (climbfuji) [reply configs/common/packages.yaml]: `+external-parallelio` is correct. I recall the issues that led to adding `~shared` and I vote for keeping it for now.

Comment (ulmononian) [reply configs/common/packages.yaml]: @AlexanderRichert-NOAA esmf variant changes made :) !

Comment (AlexanderRichert-NOAA) [reply configs/common/packages.yaml]: Cool!

**Comment by climbfuji** [comment]
CI tests passed, I am going to merge the spack PR (and then this PR needs to be updated).

**Comment by climbfuji** [comment]
I updated the submodule pointer following the merge of https://github.com/JCSDA/spack/pull/268, and reverted `.gitmodules`.

**Review by climbfuji — APPROVED** [review]

**Comment by climbfuji** [comment]
I also added the missing mapl-esmf modulefile suffices in configs/common/modules.yaml. CI tests are running again.

**Comment by climbfuji** [comment]
@AlexanderRichert-NOAA This is ready to merge once approved from your side.

**Review by srherbener — APPROVED** [review]
Thanks for updating. I tested on my M1 Mac and spack-stack built properly, as well as jedi-bundle and I saw the usual (epxected) jedi-bundle ctest failures.

**Comment by ulmononian** [comment]
> Thanks for updating. I tested on my M1 Mac and spack-stack built properly, as well as jedi-bundle and I saw the usual (epxected) jedi-bundle ctest failures.

thanks for testing steve!!

**Comment by climbfuji** [comment]
> Thanks for updating. I tested on my M1 Mac and spack-stack built properly, as well as jedi-bundle and I saw the usual (epxected) jedi-bundle ctest failures.

Thanks very much for taking the time to test this, and for the quick turnaround!

**Pull request merged**
